// @docusaurus/utils-validation/lib/validationUtils.js
"use strict";
/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateFrontMatter = exports.normalizeThemeConfig = exports.normalizePluginOptions = exports.printWarning = void 0;
const tslib_1 = require("tslib");
const logger_1 = tslib_1.__importDefault(require("@docusaurus/logger"));
const validationSchemas_1 = require("./validationSchemas");
function printWarning(warning) {
    if (warning) {
        const warningMessages = warning.details
            .map(({ message }) => message)
            .join('\n');
        logger_1.default.warn(warningMessages);
    }
}
exports.printWarning = printWarning;
function normalizePluginOptions(schema, options) {
    // All plugins can be provided an "id" option (multi-instance support)
    // we add schema validation automatically
    const finalSchema = schema.append({
        id: validationSchemas_1.PluginIdSchema,
    });
    const { error, warning, value } = finalSchema.validate(options, {
        convert: false,
    });
    printWarning(warning);
    if (error) {
        throw error;
    }
    return value;
}
exports.normalizePluginOptions = normalizePluginOptions;
function normalizeThemeConfig(schema, themeConfig) {
    // A theme should only validate his "slice" of the full themeConfig,
    // not the whole object, so we allow unknown attributes
    // otherwise one theme would fail validating the data of another theme
    const finalSchema = schema.unknown();
    const { error, warning, value } = finalSchema.validate(themeConfig, {
        convert: false,
    });
    printWarning(warning);
    if (error) {
        throw error;
    }
    return value;
}
exports.normalizeThemeConfig = normalizeThemeConfig;
function validateFrontMatter(frontMatter, schema) {
    const { value, error, warning } = schema.validate(frontMatter, {
        convert: true,
        allowUnknown: true,
        abortEarly: false,
    });
    printWarning(warning);
    if (error) {
        const frontMatterString = JSON.stringify(frontMatter, null, 2);
        const errorDetails = error.details;
        const invalidFields = errorDetails.map(({ path }) => path).join(', ');
        logger_1.default.error `The following front matter:
${logger_1.default.yellow(frontMatterString)}
contains invalid values for field(s): ${logger_1.default.yellow(invalidFields)}.
${errorDetails.map(({ message }) => message)}
`;
        throw error;
    }
    return value;
}
exports.validateFrontMatter = validateFrontMatter;
